eng/scripts/splittestdependencies/Generate-Dependency-functions.ps1

```diff
@@ -0,0 +1,72 @@
+    $projGroup = [System.Collections.ArrayList]::new()
+    for ($i = 0; $i -lt ($numOfGroups - 1); $i++) {
+      $projGroup.Add($projects[( $i * $NumberOfTestsPerJob )..(($i + 1) * $NumberOfTestsPerJob - 1)]) > $null
+    }
+    if ($projects.Count -gt ($numOfGroups - 1) * $NumberOfTestsPerJob) {
+      $projGroup.Add($projects[(($numOfGroups - 1) * $NumberOfTestsPerJob)..($projects.Count - 1)]) > $null
+    }
+    return $projGroup
+}
+
+
+# Take the group contains the project list. Write each group into a project file which can be used by dotnet test command.
+# Return the array of project files
+function Write-Test-Dependency-Group-To-Files($ProjectGroups, $MatrixOutputFolder) {
+    $null = New-Item -Path $MatrixOutputFolder -ItemType "directory" -Force 
+    $numOfGroups = $ProjectGroups.Count
+    $allOutputFiles = @()
+    for ($i = 0; $i -lt $numOfGroups; $i++) {
+      $projectFilePath = "Project_$i.props"
+      $null = New-Item -Path "$MatrixOutputFolder/$projectFilePath" -ItemType "file" -Force
+      # Retain the structure without ProjectReference.
+      $templateXml=[xml]'<Project></Project>'  
+      $projectNode = $templateXml.SelectNodes("/Project")
+      $itemGroupNode=$templateXml.CreateNode("element", "ItemGroup", "")
+      $projectNode.AppendChild($itemGroupNode) > $null
+      foreach($pkg in $ProjectGroups[$i]) {
+        $newElem = $templateXml.CreateNode("element", "ProjectReference", "")
+        $newElemAttr = $templateXml.CreateAttribute("Include")
+        $newElemAttr.InnerText = $pkg
+        $newElem.Attributes.Append($newElemAttr) > $null
+        
+        $itemGroupNode.AppendChild($newElem) > $null
+        $null = $templateXml.Save("azure-sdk-for-net/$MatrixOutputFolder/$projectFilePath")
+      }
+      $allOutputFiles += $projectFilePath
+    }
+    return $allOutputFiles
+  }
+  
+  # Add new property in the platform-matrix json and assign the values of the project file paths.
+  function Write-Project-Files-To-Matrix($ProjectFiles, $MatrixJsonPath, $MatrixOutputFolder, $ProjectFileConfigName) {
+    $platformJson = Get-Content $MatrixJsonPath | ConvertFrom-Json
+    
+    foreach ($projectFile in $ProjectFiles) {
+      $n = $projectFile -replace "Project_(\d+).props", '$1'
+      $platformJson.displayNames | Add-Member -Name $projectFile -value "DependencyGroup$n" -MemberType NoteProperty
+    }
+    
+    # Add one project files into the "include" session for short-term solution. 
+    # Assumption the include has fewer entries than project list
+    foreach ($include in $platformJson.include) {
+      $include | Add-Member -Name $ProjectFileConfigName -value $ProjectFiles[0] -MemberType NoteProperty
+      $ProjectFiles = $ProjectFiles[1..($ProjectFiles.length-1)]
+    }
+    $platformJson.matrix | Add-Member -Name $ProjectFileConfigName -value $ProjectFiles -MemberType NoteProperty
+    $platformJson | ConvertTo-Json -Depth 100 | Out-File "$MatrixOutputFolder/platform-matrix.json" 
```

Comment:
The jobs from the include section show up at the end of the matrix, so I think it would be logical to have the test project slice also come from the end so that the slice ordering aligns with the matrix job ordering. This will help people to reason from the config deterministically about which tests will run on which platforms.

It will slightly complicate the code, but maybe something like:

```suggestion
    $matrixSlice = $ProjectFiles | Select -SkipLast $platformJson.include.Length
    $includeSlice = $ProjectFiles | Select -Skip $matrixSlice.Length
    for ($i = 0; $i -lt $includeSlices.Length; $i++) {
      $platformJson.include[$i] | Add-Member -Name $ProjectFileConfigName -value $includeSlice[$i] -MemberType NoteProperty
    }
    $platformJson.matrix | Add-Member -Name $ProjectFileConfigName -value $matrixSlice -MemberType NoteProperty
    $platformJson | ConvertTo-Json -Depth 100 | Out-File "$MatrixOutputFolder/platform-matrix.json" 
```

Reply:
This is a better way to pop the first. 

The structure changed as we decided to have the property file as a new json property, so we can have more values for the each file. 
Here is the new look:
<img width="339" alt="image" src="https://user-images.githubusercontent.com/48036328/183141206-4c46ff5c-943d-44ea-b721-6c7b809a8c31.png">
 Now it turns to pop the first property from the object

Reply:
In your above example, I would not want `Project_0.props` to be the override file for the include, for example if the matrix had three test entries, then the ordering of tests would be unintuitive:

```
Project_1.props
Project_2.props
Project_3.props
Project_0.props
```